### ISSUE TITLE
Switching between implementation or test in project specific manner.

### ISSUE THREAD

**Issue by proofit404** [bbatsov/projectile#268]
Currently projectile search corresponding tests by prefixes and suffices. This approach isn't good in general. Some projects (Django for example) use different naming approach. It just a `tests.py` file in same directory with view, controller and model. Projectile already determine project manner for compiling project and run its tests. I think this issue is the same.

**Comment by hynek** [comment]
Just to pile on my use case that doesn’t work either currently: I have my Python source code in an src directory and my tests in a tests directory and projectile can’t find each other even though they have same names (e.g. `src/project/utils.py` and `tests/test_utils.py`).

I’m not sure what exactly makes it not work but it would be nice if you work on this feature if you could take my use case into consideration. :)

**Comment by omajid** [comment]
One more use case that doesn't work:

I have an [autotools-based python project]( https://fedorahosted.org/system-switch-java/browser)with a layout roughly like:
- `Makefile`
- `switch_java_functions.py`
- `tests/test_switch_java_functions.py`

Because the root is a `Makefile`, not a `setup.py`, projectile thinks this is a `make` project and in the `test_switch_java_functions.py` file, `(projectile-test-file-p (buffer-file-name))` evaluates to `nil`.

**Comment by dwang20151005** [comment]
Another use-case from my work, and a proposal:

In my work, the important pair I jump between is not implementation and test but
implementation and interface, and for unimportant reasons the implementation
foo.x can have its interface either in `foo.xi` or in `foo_intf.x`. But it's
still true for my work that you can look at a file and predict exactly the names
of the other files.

My proposal is like what you sketched: replace `projectile-other-file-alist`
with a (per-project) variable `projectile-other-file-function` s.t. `(funcall
projectile-other-file-function current-path flex-matching)` returns a list of
regexes matching related paths, respecting `flex-matching`. Both `current-path`
and the returned regexes are relative to the project root.

This is a drop-in generalization of how you currently populate `file-list` in
`projectile-get-other-files`. It handles n-way relations, like my silly
interface names, and tests located elsewhere in the tree, like hynek's and
omajid's, and tests whose names don't contain the current path, like
proofit404's.

To ease the transition, we could provide known project types with default values
of `projectile-other-file-function` that recapitulate the current prefix/suffix
behavior.

I'm really, really not an elisp programmer, so it's questionable whether
offering a pull request would actually be polite, but I do make the offer.

**Comment by evgeniysharapov** [comment]
There's already `projectile-test-suffix/prefix-function` that could be used to check whether file is a test file. Perhaps instead of one custom value it should be on a per project basis (which could be done at the moment with dir locals).  We could allow registration of a project based test file preficate `projectile-register-project-type` function.

**Comment by DamienCassou** [comment]
Adding functions to `.dir-locals.el` is sub-optimal in my opinion as I believe that triggers unsafe variable warnings each time you open a project's file.

**Comment by justinabrahms** [comment]
I'm running into this today.

I've tried hacking around in `projectile.el`, but am finding that there's a fair amount of duplication about how tests are discovered.

As an example, `projectile-create-test-file-for` builds the test file from the implementation, whereas `projectile-find-matching-test-file` doesn't consider this building of test names at all. It would be ideal (for me, at least), if we attempted to build test names and just go to them based on the find/replace of the implementation file.

Example:

If I'm in `src/libraries/db.js`, I'd like the test-finding mechanisms to try `test/libraries/db.js` first (which is how my project is setup via `:test-dir`, `:src-dir`, and an empty `:test-suffix`. If it doesn't find that file.. maybe then it goes searching.. but that's less useful to me.

I also ran into a few other problems:

1. If your source dir is `Src` and your test directory it `test`, the test file it tries to make is `Test` because the find/replace on the path doesn't preserve case.
1. When looking for implementations to swap to for the test->src direction, things in `test-dir` aren't excluded from the candidates.

Is this bubbling up to the top of anyone's todos in the near term, by chance?
